src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -93,30 +152,162 @@ void GenericOTARequestorDriver::UpdateSuspended(System::Clock::Seconds32 delay)
+}
+
+void GenericOTARequestorDriver::CancelDelayedAction(System::TimerCompleteCallback action, void * aAppState)
+{
+    SystemLayer().CancelTimer(action, aAppState);
 }
 
-void GenericOTARequestorDriver::ScheduleDelayedAction(UpdateFailureState state, System::Clock::Seconds32 delay,
-                                                      System::TimerCompleteCallback action)
+void GenericOTARequestorDriver::ProcessAnnounceOTAProviders(
+    const ProviderLocationType & providerLocation,
+    app::Clusters::OtaSoftwareUpdateRequestor::OTAAnnouncementReason announcementReason)
 {
-    CHIP_ERROR error = SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, this);
+    // If reason is URGENT_UPDATE_AVAILABLE, we start OTA immediately. Otherwise, respect the timer value set in mOtaStartDelayMs.
+    // This is done to exemplify what a real-world OTA Requestor might do while also being configurable enough to use as a test app.
+    uint32_t msToStart = 0;
+    switch (announcementReason)
+    {
+    case OTAAnnouncementReason::kSimpleAnnouncement:
+    case OTAAnnouncementReason::kUpdateAvailable:
+        msToStart = mOtaStartDelayMs;
+        break;
+    case OTAAnnouncementReason::kUrgentUpdateAvailable:
+        msToStart = kImmediateStartDelayMs;
+        break;
+    default:
+        ChipLogError(SoftwareUpdate, "Unexpected announcementReason: %u", static_cast<uint8_t>(announcementReason));
+        return;
+    }
+
+    // IMPLEMENTATION CHOICE:
+    // This implementation of the OTARequestor driver ignores the announcement if an update is in progress,
+    // otherwise it queries the provider passed in the announcement
 
-    if (error != CHIP_NO_ERROR)
+    if (mRequestor->GetCurrentUpdateState() != OTAUpdateStateEnum::kIdle)
     {
-        HandleError(state, error);
+        ChipLogProgress(SoftwareUpdate, "State is not kIdle, ignoring the AnnounceOTAProviders. State: %d",
+                        (int) mRequestor->GetCurrentUpdateState());
+        return;
     }
+
+    // Point the OTARequestor to the announced provider
+    mRequestor->SetCurrentProviderLocation(providerLocation);
+    mLastUsedProvider = providerLocation;
+
+    ScheduleDelayedAction(System::Clock::Seconds32(msToStart / 1000), StartDelayTimerHandler, this);
+}
+
+void GenericOTARequestorDriver::DriverSendQuery()
+{
+
+    // IMPLEMENTATION CHOICE
+    // In this implementation explicitly triggering a query cancels any in-progress update.
+    UpdateCancelled();
+
+    // Default provider timer only runs when there is no ongoing query/update; must stop it now.
+    // ConnectToProvider() will change the state from kIdle
+    StopDefaultProviderTimer();
+
+    mRequestor->ConnectToProvider(OTARequestorInterface::kQueryImage);
+}
+
+void GenericOTARequestorDriver::DefaultProviderTimerHandler(System::Layer * systemLayer, void * appState)
```

Comment:
It feels like this function and `StartDelayTimerHandler` can be combined? The only real difference is one looks up the default OTA provider to use and the other doesn't. It seems we could use the `appState` param to differentiate those two?

Reply:
Based on suggestion https://github.com/project-chip/connectedhomeip/pull/15433/files#r814191405, then maybe we can't combine the two functions.

Reply:
Yes, we cannot.